unit/suites/test_mesh.py

```diff
@@ -0,0 +1,107 @@
+        self.srv.start()
+        self.srv.admin("box.schema.user.create('test', { password = 'test', if_not_exists = true })")
+        self.srv.admin("box.schema.user.grant('test', 'execute', 'universe')")
+        
+        # Start second server
+        self.srv2 = TarantoolServer()
+        self.srv2.script = 'unit/suites/box.lua'
+        self.srv2.start()
+        self.srv2.admin("box.schema.user.create('test', { password = 'test', if_not_exists = true })")
+        self.srv2.admin("box.schema.user.grant('test', 'execute', 'universe')")
+
+        # get_nodes function contains both servers' addresses
+        get_nodes = " \
+            function get_cluster_nodes() \
+                return { '%s:%d', '%s:%d' } \
+            end" % (self.srv.host, self.srv.args['primary'], self.srv2.host, self.srv2.args['primary'])
+
+        # Create get_nodes function on servers
+        self.srv.admin(get_nodes)
+        self.srv2.admin(get_nodes)
+
+        # Create srv_id function (for testing purposes)
+        self.srv.admin("function srv_id() return 1 end")
+        self.srv2.admin("function srv_id() return 2 end")
+
+
+    def setUp(self):
+        # prevent a remote tarantool from clean our session
+        if self.srv.is_started():
+            self.srv.touch_lock()
+        
+        if self.srv2.is_started():
+            self.srv2.touch_lock()
+
+    def test_01_connect_and_refresh(self):
+
+        # Create a mesh connection, pass only the first server address
+        con = tarantool.MeshConnection(self.srv.host, self.srv.args['primary'],
+            user='test',
+            password='test',
+            get_nodes_function_name='get_cluster_nodes',
+            connect_now=True)
+
+        # Check strategy have one addr from config
+        self.assertIs(len(con.strategy.addrs), 1) 
+
+        con.last_nodes_refresh = 0
+
+        # Check we work with the first server
+        resp = con.call('srv_id')
+        self.assertIs(resp.data and resp.data[0] == 1, True)
+        
+        # Check refresh is successful and strategy have 2 nodes
+        self.assertIs(len(con.strategy.addrs), 2)
+
+        con.close()
+
+    def test_02_mesh_exclude_node(self):
+        # Create a mesh connection, pass only the first server address
+        con = tarantool.MeshConnection(self.srv.host, self.srv.args['primary'],
+            user='test',
+            password='test',
+            get_nodes_function_name='get_cluster_nodes',
+            connect_now=True)
+
+        con.last_nodes_refresh = 0
+        resp = con.call('srv_id')
+        self.assertIs(resp.data and resp.data[0] == 1, True)
+        
+        # Check refresh is successful and strategy have 2 nodes
+        self.assertIs(len(con.strategy.addrs), 2)
+        self.srv.stop()
+        
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")
+            resp = con.call('srv_id')
+        
+        self.assertIs(resp.data and resp.data[0] == 2, True)
```

Comment:
I would also check here that the strategy contains only one URL.

Reply:
Ouch, no, it still has two addresses. This is the test for reconnection, not for cluster discovery. We have no test, where an instance is excluded from a list, because a cluster discovery function returns a list w/o it. We also have no test for improper function answer (say, when an error is risen). There is no test, which pass `addrs` and also `host` and `port` simultaneously.

Mix of these things (reconnection and cluster discovery I mean) in tests suggests me again that we should perform reconnection in MeshConnection._opt_reconnect() and doing cluster discovery in MeshConnection._send_request() (or _before_send_request()). This will give us more clear vision how should test scenarious look like.

Also I suggest to look into related tarantool-java test cases:

Re cluster client behaviour: https://github.com/tarantool/tarantool-java/blob/762ce2fff44bd08b95358bfed850b3fc15423fa3/src/test/java/org/tarantool/ClientReconnectClusterIT.java
Re validation errors: https://github.com/tarantool/tarantool-java/blob/762ce2fff44bd08b95358bfed850b3fc15423fa3/src/test/java/org/tarantool/cluster/ClusterServiceStoredFunctionDiscovererIT.java

Reply:
Tests reworked. Added test for configuration.